### DISCUSSION TITLE
Use of Finite Strain with Tensor Mechanics New System

### DISCUSSION THREAD

**Discussion by Edward-Eth** [idaholab/moose#22784] [Q&A Modules: Solid mechanics] [answered]
I've been attempting to use the Tensor Mechanics New System, as some brief investigation suggested to me that it provided better convergence for finite strain than the old system, but I'm now uncertain that my implementation was correct. Previously I was using this as my input:

```
[Modules/TensorMechanics/Master]
  [all]
    strain = FINITE
    new_system = true
    incremental = true
    add_variables = true
    generate_output = 'l2norm_strain'
  []
[]

[Materials]
  [Elasticity]
    type = ComputeIsotropicElasticityTensor
    youngs_modulus = 110000
    shear_modulus = 42000
    use_displaced_mesh = true
  []
  [Stress]
    type = ComputeLagrangianLinearElasticStress
  []
[]
```

Which was offering far superior convergence to the same options on the old system (with ComputeLagrangianLinearElasticStress replaced with ComputeLinearElasticStress). However I have since done more digging and discovered the "large_kinematics" flag. Is this the "true" toggle for FINITE strain in the new system or am I misunderstanding it? When I add large_kinematics = True to the stress material with the new system I invariably get non-convergence and negative Jacobians for a simple tip displaced (dirichlet) beam problem.

**Comment by hugary1995** [comment, marked as answer]
Yes, to turn on large deformation in the new system you need to do the following
1. Set `large_kinematics = true` in the `[GlobalParams]`
2. Set `new_system = true` and `strain = FINITE` in the TM action.

The `incremental` parameter in the TM action is ignored if `new_system = true`.
If you want to use total Lagrangian just set `formulation = TOTAL`.
If you want to use updated Lagrangian set `formulation = UPDATED`.

We spent a significant amount of effort to get the Jacobian of the new system exact. So if you encounter convergence issues with the new system, feel free to send us the minimal reproducing example.

**Reply by Edward-Eth** [reply]
When I use those parameters, with the following input:
```
[Mesh]
  [Generated]
    type = GeneratedMeshGenerator
    dim = 3
    elem_type = HEX8
    xmax = 500
    ymax = 25
    zmax = 25
    nx = 100
    ny = 5
    nz = 5
  []
[]

[GlobalParams]
  displacements = 'disp_x disp_y disp_z'
  large_kinematics = true
[]

[Modules/TensorMechanics/Master]
  [all]
    strain = FINITE
    new_system = true
    add_variables = true
  []
[]

[Materials]
  [Elasticity]
    type = ComputeIsotropicElasticityTensor
    youngs_modulus = 110000
    shear_modulus = 42000
    use_displaced_mesh = true
  []
  [Stress]
    type = ComputeLagrangianLinearElasticStress
  []
[]

[BCs]
  [Fixed_X]
    type = DirichletBC
    boundary = 'left'
    value = 0
    variable = disp_x
    use_displaced_mesh = true
  []
  [Fixed_Y]
    type = DirichletBC
    boundary = 'left'
    value = 0
    variable = disp_y
    use_displaced_mesh = true
  []
  [Fixed_Z]
    type = DirichletBC
    boundary = 'left'
    value = 0
    variable = disp_z
    use_displaced_mesh = true
  []
  [Tip_Displacement]
    type = DirichletBC
    boundary = 'right'
    variable = disp_z
    value = 50
    use_displaced_mesh = true
  []
[]

[Executioner]
  type = Steady

  automatic_scaling = true
  solve_type = 'NEWTON'

  petsc_options_iname = '-pc_type  -pc_hypre_type'
  petsc_options_value = 'hypre     boomeramg'
[]

[Outputs]
  [Performance]
    type = PerfGraphOutput
  []
[]
```
I get:
```
ERROR: negative Jacobian -0.589091 at point (x,y,z)=( 490.063,  9.73302,   56.166) in element 1197
[6] ../src/fe/fe_map.C, line 1178, compiled Oct  5 2022 at 23:16:04
We caught a libMesh error in ThreadedElementLoopBase
  Elem Information
   id()=2096, unique_id()=5732, processor_id()=6
   type()=HEX8
   dim()=3
   n_nodes()=8
   mapping=LAGRANGE_MAP
    0  Node id()=2616, processor_id()=6, Point=(x,y,z)=( 480.766,  0.48683,   57.037)
    DoFs=(0/0/9228) (0/1/9229) (0/2/9230) 
    1  Node id()=2618, processor_id()=6, Point=(x,y,z)=( 485.842,  1.21273,  59.5806)
    DoFs=(0/0/9234) (0/1/9235) (0/2/9236) 
    2  Node id()=2619, processor_id()=6, Point=(x,y,z)=( 485.823,  5.45591,  62.0315)
    DoFs=(0/0/9237) (0/1/9238) (0/2/9239) 
    3  Node id()=2617, processor_id()=6, Point=(x,y,z)=( 480.649,  5.25947,  57.9656)
    DoFs=(0/0/9231) (0/1/9232) (0/2/9233) 
    4  Node id()=3222, processor_id()=6, Point=(x,y,z)=( 480.847, -1.39579,  62.2696)
    DoFs=(0/0/9480) (0/1/9481) (0/2/9482) 
    5  Node id()=3224, processor_id()=6, Point=(x,y,z)=( 485.933,  4.97892,  61.1484)
    DoFs=(0/0/9486) (0/1/9487) (0/2/9488) 
    6  Node id()=3225, processor_id()=6, Point=(x,y,z)=( 485.423,  3.60393,  63.1854)
    DoFs=(0/0/9489) (0/1/9490) (0/2/9491) 
    7  Node id()=3223, processor_id()=6, Point=(x,y,z)=(  480.42,  4.60968,  63.2469)
    DoFs=(0/0/9483) (0/1/9484) (0/2/9485) 
   n_sides()=6
    neighbor(0)=1596
    neighbor(1)=nullptr
    neighbor(2)=2097
    neighbor(3)=2196
    neighbor(4)=2095
    neighbor(5)=nullptr
   hmin()=1.00977, hmax()=8.67242
   volume()=66.4199
   active()=1, ancestor()=0, subactive()=0, has_children()=0
   parent()=nullptr
   level()=0, p_level()=0
   refinement_flag()=DO_NOTHING
   p_refinement_flag()=DO_NOTHING
   DoFs=
   ```
   ETC

**Reply by Edward-Eth** [reply]
Whereas the same file runs fine without large kinematics and with strain = SMALL

**Reply by Edward-Eth** [reply]
If I split the issue down into at least 4 steps it can solve, but still requires a large number of iterations. Is this an unavoidable issue with the finite strain implementation?

EDIT:
It was working on 4 steps:
```
[Mesh]
  [Generated]
    type = GeneratedMeshGenerator
    dim = 3
    elem_type = HEX8
    xmax = 500
    ymax = 25
    zmax = 25
    nx = 100
    ny = 5
    nz = 5
  []
[]

[GlobalParams]
  displacements = 'disp_x disp_y disp_z'
  large_kinematics = true
[]

[Modules/TensorMechanics/Master]
  [all]
    strain = FINITE
    new_system = true
    add_variables = true
  []
[]

[Materials]
  [Elasticity]
    type = ComputeIsotropicElasticityTensor
    youngs_modulus = 110000
    shear_modulus = 42000
    use_displaced_mesh = true
  []
  [Stress]
    type = ComputeLagrangianLinearElasticStress
  []
[]

[BCs]
  [Fixed_X]
    type = DirichletBC
    boundary = 'left'
    value = 0
    variable = disp_x
    use_displaced_mesh = true
  []
  [Fixed_Y]
    type = DirichletBC
    boundary = 'left'
    value = 0
    variable = disp_y
    use_displaced_mesh = true
  []
  [Fixed_Z]
    type = DirichletBC
    boundary = 'left'
    value = 0
    variable = disp_z
    use_displaced_mesh = true
  []
  [Tip_Displacement]
    type = FunctionDirichletBC
    boundary = 'right'
    variable = disp_z
    function = 50*t
    use_displaced_mesh = true
  []
[]

[Executioner]
  type = Transient

  dt = 0.25

  end_time = 1.0

  automatic_scaling = true
  solve_type = 'NEWTON'

  petsc_options_iname = '-pc_type  -pc_hypre_type'
  petsc_options_value = 'hypre     boomeramg'
[]

[Outputs]
  [Performance]
    type = PerfGraphOutput
  []
[]
```
However if I refine the mesh to 200 x 10 x 10 it once again suffers negative jacobians.

**Reply by hugary1995** [reply]
I just ran your input on my laptop. The behavior makes sense to me. A few considerations:
- Applying 10% strain in one step is not going to work with any geometrically nonlinear implementation, so you definitely need to split it into many steps. In fact, splitting into 4 steps with each step incrementing 2.5% strain still sounds a lot. Typically, you don't want to increase the strain more than 1% per step.
- It's not using "a lot of iterations" as I see it. It uses at most 7 nonlinear iterations. Again, this is because you are posing a large strain increment per step.
- Hypre/boomeramg is a algebraic multigrid solver. It's iterative in nature, so you will see many linear iterations. It also has a ton of parameters to tune, and I am sure the model can converge much better if you spend some time tuning the solver parameters. As always, if you use a direct solver, it will converge in 1 linear iteration, but the scaling efficiency won't be as good.
- I made some modifications to your input file. Namely increasing the number of steps to 10 and adding a simple predictor. The model takes 7 nl itrs to solve in the first step, and 3 nl itrs in the subsequent steps. Note that this is still with the default settings of hypre/boomeramg.

```python
[Mesh]
  [Generated]
    type = GeneratedMeshGenerator
    dim = 3
    elem_type = HEX8
    xmax = 500
    ymax = 25
    zmax = 25
    nx = 100
    ny = 5
    nz = 5
  []
[]

[GlobalParams]
  displacements = 'disp_x disp_y disp_z'
  large_kinematics = true
[]

[Modules/TensorMechanics/Master]
  [all]
    strain = FINITE
    new_system = true
    add_variables = true
  []
[]

[Materials]
  [Elasticity]
    type = ComputeIsotropicElasticityTensor
    youngs_modulus = 110000
    shear_modulus = 42000
    use_displaced_mesh = true
  []
  [Stress]
    type = ComputeLagrangianLinearElasticStress
  []
[]

[BCs]
  [Fixed_X]
    type = DirichletBC
    boundary = 'left'
    value = 0
    variable = disp_x
    use_displaced_mesh = true
  []
  [Fixed_Y]
    type = DirichletBC
    boundary = 'left'
    value = 0
    variable = disp_y
    use_displaced_mesh = true
  []
  [Fixed_Z]
    type = DirichletBC
    boundary = 'left'
    value = 0
    variable = disp_z
    use_displaced_mesh = true
  []
  [Tip_Displacement]
    type = FunctionDirichletBC
    boundary = 'right'
    variable = disp_z
    function = 50*t
    use_displaced_mesh = true
  []
[]

[Executioner]
  type = Transient

  dt = 0.1

  end_time = 1.0

  automatic_scaling = true
  solve_type = 'NEWTON'

  petsc_options_iname = '-pc_type  -pc_hypre_type'
  petsc_options_value = 'hypre     boomeramg'

  [Predictor]
    type = SimplePredictor
    scale = 1
  []
[]

[Outputs]
  [Performance]
    type = PerfGraphOutput
  []
  print_linear_residuals = false
[]
```

**Reply by hugary1995** [reply]
Another important consideration: the hypoelasticity solution (what you are using here) only converges to the ground-truth analytical solution as dt -> 0. So if you use a large time step, not only it will suffer from numerical convergence issues in newton-raphson iterations, but also you are sacrificing the accuracy of the solution.

As an alternative, we also provide some predefined hyperelasticity models, e.g. St. Venant Kirchhoff, Neo-Hookean, etc., which don't have time integration errors. You can also implement your own favorite hyperelastic material.

**Reply by Edward-Eth** [reply]
With the changes you have provided the model does work better. You imply that the issue is to do with Newton solving, do you think PJFNK might offer better stability in this regard? The simulations I intend to perform have long time scales so being limited to ensure less than 1% strain occurs per step would lead to very large step counts, and would be hard to enforce/predict correctly.

I am aware of the benefits of the direct solver (lu) but with the size of the models I wish to run it's not a viable option due to memory limits.

I am also unsure of my implementation of use_displaced_mesh in this problem case, as I have seen some MOOSE discussion posts suggesting it is best included and others suggesting it is best omitted. It's not clear to me the physical meaning of applying a use_displaced_mesh = true to a fixed dirichlet condition for example.

On the topic of hyperelastic models, I am not clear on how these can be applied to metals with typically linear elastic behaviour up to plastic deformation (which is something I have yet to get figured out in MOOSE but a separate topic)

**Reply by hugary1995** [reply]
> With the changes you have provided the model does work better. You imply that the issue is to do with Newton solving, do you think PJFNK might offer better stability in this regard? 

If the Jacobian is exact, I recommend Newton over PJFNK.

> The simulations I intend to perform have long time scales so being limited to ensure less than 1% strain occurs per step would lead to very large step counts, and would be hard to enforce/predict correctly.

"long time scale" would imply that the time step size you might want to use is large. However, "large time step size" does not necessarily correspond to "large strain increment". Does that make sense? For example, you may need 1e6 time steps with each step of size 1000 seconds, but the strain increment may still be small. In other words, a problem with a "long time scale" isn't well-posed if say you need 1e6 time steps, and each step you need to increase strain by 5% -- the body would explode.

> I am aware of the benefits of the direct solver (lu) but with the size of the models I wish to run it's not a viable option due to memory limits.

Understood.

> I am also unsure of my implementation of use_displaced_mesh in this problem case, as I have seen some MOOSE discussion posts suggesting it is best included and others suggesting it is best omitted. It's not clear to me the physical meaning of applying a use_displaced_mesh = true to a fixed dirichlet condition for example.

Yeah, this is probably the single-most confusing concept for new users. I feel like we should have a specific page dedicated to explaining this...

Anyways, a displaced mesh is mapped from the initial mesh by applying the displacements. If no displacements is applied to the mesh, then the displaced mesh will fall back to the initial mesh.

When `use_displaced_mesh = false`, everything _this_ object evaluates are computed on the initial mesh. Otherwise if `use_displaced_mesh = true`, everything is evaluated on the displaced mesh. TL;DR

So obviously `use_displaced_mesh` only has an effect if the quantities evaluate to different values on the displaced mesh and the initial mesh. For example,
1. Quadrature points in the parametric coordinates remain the same;
2. Physical coordinates of quadrature points and nodes are _different_ on displaced/initial mesh;
3. Due to 1, test and shape functions evaluate to the same value regardless of displaced/initial mesh at quadrature points and nodes, as they are defined in terms of the parametric coordinates;
4. Due to 2, gradients of test and shape functions are _different_ on displaced/initial mesh, since the isoparametric mapping depends on the physical coordinates of the nodes;
5. Due to 3, variable values at quadrature points and nodes remain unchanged on displaced/initial mesh, for instance the displacements, temperature, etc.;
6. Due to 4, variable gradients at quadrature points are different on displaced/initial mesh, for instance the strain, stress, etc., (hence why we always insist on strain calculators run on the initial undisplaced mesh);
7. Constant material properties are not affected by the mesh, for instance your `ComputeIsotropicElasticityTensor`;
8. In contrast to 7, if the material property depends on the spatial coordinates, its value will be different on displaced/initial mesh;
9. The same node in the displaced/initial mesh will correspond to the same degree of freedom in the system. A Dirichlet (nodal) BC is not affected by `use_displaced_mesh` if and only if the value you are setting does not depend on the mesh configuration. For instance, the constant `DirichletBC` in your input file does not care about `use_displaced_mesh`.

> On the topic of hyperelastic models, I am not clear on how these can be applied to metals with typically linear elastic behaviour up to plastic deformation (which is something I have yet to get figured out in MOOSE but a separate topic)

We also have a predefined Simo-Hughes style hyperelastic-plastic constitutive model in the module.

**Reply by Edward-Eth** [reply]
This is all extremely helpful, thank you. 

I will look at the Simo-Hughes model, it seems like it might be what I need to use (although I'm not 100% clear on how I'd implement temperature dependant yield with it but again, an issue for a different day).

Continuing on the topic of displaced mesh, I have a few more questions:
1) In the case where I have a temperature dependent elasticity tensor, but temperature is defined at nodes not spatially, I assume this means that use_displaced_mesh will/should have no effect and therefore should be false
2) In cases such as a radiative boundary condition, use_displaced_mesh should be true if I expect the boundary to change shape/size during the simulation
3) In heat conduction, use_displaced_mesh will obviously change the thermal behaviour if displacement is occurring, as lengths between nodes are changing for example, however I have a question: 
   - With a ConstantGenericMaterial defining my density, it is a spatially and thermally invariant value. 
   - I then use this property in a model which undergoes significant deformation (say stretches to 2x its initial length)
   - If my poisson's ratio is not 0.5 (ie not incompressible) there will be some change in the volume of my model
   - I will want to use_displaced_mesh in HeatConduction and TimeDerivative in this case, as the lengths and areas will have changed significantly, altering the conductive behaviour
   - If I use_displaced_mesh in HeatConduction and HeatConductionTimeDerivative, will my effective mass be changed, due to an increased volume but unchanged density?
   - If so, are there any fixes for this?

**Reply by Edward-Eth** [reply]
Sorry, really picking your brains here now. I've just discovered that changing preset to false in my tip displacement boundary condition makes a huge difference to convergence. I think I understand this but wanted to run my hypothesis by someone who actually knows what they're doing before running with it 😆 

My understanding of preset/not preset BCs is that when preset, they will be used in forming the first nonlinear iteration, Pre-conditioner and Jacobian. When not preset they are omitted. My hypothesis of why not presetting the displacement improves my simulation is that it eliminates the initial high skew applied to the tip elements when making the Pre-conditioners etc, giving a better initial solve condition than is produced with the inclusion of the skew. Does this hypothesis make any sense to you?

I got the idea to set preset to false from the example input file on the Simo-Hughes J2 plasticity material you suggested I look at.

**Reply by hugary1995** [reply]
Sorry for the late response -- coming back from the holiday.

> This is all extremely helpful, thank you.
> 
> I will look at the Simo-Hughes model, it seems like it might be what I need to use (although I'm not 100% clear on how I'd implement temperature dependant yield with it but again, an issue for a different day).
> 
> Continuing on the topic of displaced mesh, I have a few more questions:
> 
> 1. In the case where I have a temperature dependent elasticity tensor, but temperature is defined at nodes not spatially, I assume this means that use_displaced_mesh will/should have no effect and therefore should be false

Correct. Either true or false.

> 2. In cases such as a radiative boundary condition, use_displaced_mesh should be true if I expect the boundary to change shape/size during the simulation

Correct. Especially if you are computing the view factors.

> 3. In heat conduction, use_displaced_mesh will obviously change the thermal behaviour if displacement is occurring, as lengths between nodes are changing for example, however I have a question:
>    
>    * With a ConstantGenericMaterial defining my density, it is a spatially and thermally invariant value.
>    * I then use this property in a model which undergoes significant deformation (say stretches to 2x its initial length)
>    * If my poisson's ratio is not 0.5 (ie not incompressible) there will be some change in the volume of my model
>    * I will want to use_displaced_mesh in HeatConduction and TimeDerivative in this case, as the lengths and areas will have changed significantly, altering the conductive behaviour
>    * If I use_displaced_mesh in HeatConduction and HeatConductionTimeDerivative, will my effective mass be changed, due to an increased volume but unchanged density?

This thread may be helpful: https://github.com/idaholab/moose/discussions/22467#discussioncomment-3937986
Take a look and let me know which case of 1~3 you were referring to, and we can continue our discussion here.
>    * If so, are there any fixes for this?

**Reply by hugary1995** [reply]
> Sorry, really picking your brains here now. I've just discovered that changing preset to false in my tip displacement boundary condition makes a huge difference to convergence. I think I understand this but wanted to run my hypothesis by someone who actually knows what they're doing before running with it 😆
> 
> My understanding of preset/not preset BCs is that when preset, they will be used in forming the first nonlinear iteration, Pre-conditioner and Jacobian. When not preset they are omitted. My hypothesis of why not presetting the displacement improves my simulation is that it eliminates the initial high skew applied to the tip elements when making the Pre-conditioners etc, giving a better initial solve condition than is produced with the inclusion of the skew. Does this hypothesis make any sense to you?
> 
> I got the idea to set preset to false from the example input file on the Simo-Hughes J2 plasticity material you suggested I look at.

Yes. I mentioned this in another thread: https://github.com/idaholab/moose/discussions/22348#discussioncomment-3881003

**Reply by Edward-Eth** [reply]
No worries about the slow response, hope you had a good thanksgiving/holiday

I think using Density material properties accounts for the issue I was talking about, so that's one concern ticked off. 

Also glad to hear I'm not totally spouting nonsense with the preset BC issue, I'll keep that in mind in future.

And thanks again for greatly enhancing my understanding of the displaced_mesh system, that'll be very helpful :)
